src/common/core/corediff.js

```diff
@@ -811,6 +813,89 @@ define(['common/util/canon',
+            //there is three types of path equality:
+            //1. same guids -> same node
+            //2. both was moved -> different nodes
+            //3. one was moved and the other is created ->different nodes (here we always have to generate
+            // new relid to the moved one)
+            var i,
+                keys = getDiffChildrenRelids(diffBase),
+                globalDiff,
+                newRelid,
+                newPath,
+                nodeDiff,
+                relids,
+                dst2src,
+                src2dst,
+                relidObj = {},
+                parent,
+                needToRelocate = false;
+
+            if (diffBase.guid !== diffExtension.guid) {
+                if (diffBase.movedFrom && diffExtension.movedFrom) {
+                    //relocate the extension
+                    globalDiff = _concatExtension;
+                    nodeDiff = diffExtension;
+                    dst2src = _concatMoves.getExtensionSourceFromDestination;
+                    src2dst = _concatMoves.getExtensionDestinationFromSource;
+                    needToRelocate = true;
+                } else if (diffBase.movedFrom && diffExtension.removed === false) {
+                    globalDiff = _concatBase;
+                    nodeDiff = diffBase;
+                    dst2src = _concatMoves.getBaseSourceFromDestination;
+                    src2dst = _concatMoves.getBaseDestinationFromSource;
+                    needToRelocate = true;
+                } else if (diffExtension.movedFrom && diffBase.removed === false) {
+                    globalDiff = _concatExtension;
+                    nodeDiff = diffExtension;
+                    dst2src = _concatMoves.getExtensionSourceFromDestination;
+                    src2dst = _concatMoves.getExtensionDestinationFromSource;
+                    needToRelocate = true;
+                }
+
+                if (needToRelocate) {
+                    relids = getDiffChildrenRelids(getPathOfDiff(_concatBase, getParentPath(path)))
+                        .concat(getDiffChildrenRelids(getPathOfDiff(_concatExtension, getParentPath(path))));
+
+                    relidObj = {};
+                    for (i = 0; i < relids.length; i += 1) {
+                        relidObj[relids[i]] = {};
+                    }
+                    newRelid = RANDOM.generateRelid(relidObj);
+                    newPath = getParentPath(path) + '/' + newRelid;
+
+                    //now the actual place switching
+                    parent = getPathOfDiff(globalDiff, getParentPath(path));
+                    parent[newRelid] = nodeDiff;
+                    delete parent[relid];
+                    dst2src[newPath] = dst2src[path];
+                    delete dst2src[path];
+                    src2dst[dst2src[newPath]] = newPath;
+                } else {
+                    logger.warning('there is a guid mismatch among the two diffs: ' +
```

Comment:
Is this the unexpected, what are the consequences? Maybe throw an error in an else at 863?